### PR TITLE
Added support for models on AWS Bedrock and Sagemaker

### PR DESCRIPTION
**Pull request by drawal1**
Supported model families: Mistral, Anthropic, Meta

This replaced PR #772 which is closed. Code has been updated to satisfy review comments on PR #772 posted by @arnavsinghvi11

### CODE REVIEW

**Comment by drawal1** [comment]
@arnavsinghvi11 - Thanks for the review! I have updated all the code per your review comments. 

Notes
1. I have deleted the old bedrock.py since it is obsolete and the name would conflict with the new aws provider class called Bedrock
2. There is a test_aws_models.py file under /tests/modules which does not test fixtures for aws credentials etc. so its not ready for pytest yet, but it is a nice, comprehensive test for all the aws models and providers and also illustrates a helper function that makes it easy to use aws providers and models

**Comment by drawal1** [comment]
@sachinkun21 - can you get the latest and doublecheck that aws_models.py contains below code?

```
        # Anthropic models do not support the following parameters
        query_args.pop("frequency_penalty", None)
        query_args.pop("num_generations", None)
        query_args.pop("presence_penalty", None)
        query_args.pop("model", None)
```

File: docs/api/language_model_clients/aws.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+### Usage
+
+```python
+# Notes:
+# 1. Install boto3 to use AWS models.
+# 2. Configure your AWS credentials with the AWS CLI before using these models
+
+# initialize the bedrock aws provider
+bedrock = Bedrock(region_name="us-west-2")
```
Comment (arnavsinghvi11) [inline docs/api/language_model_clients/aws.md:15]: the LM calls here should be `dspy.Bedrock`, etc.

File: docs/api/language_model_clients/aws.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+sagemaker = Sagemaker(region_name="us-west-2")
+# For mistral on Sagemaker
+# Note: you need to create a Sagemaker endpoint for the mistral model first
+lm = AWSMistral(sagemaker, "<YOUR_MISTRAL_ENDPOINT_NAME>", **kwargs)
+
+```
+
+### Constructor
+
+The constructor initializes the base class `LM` and the `AWSProvider` class.
```
Comment (arnavsinghvi11) [inline docs/api/language_model_clients/aws.md:33]: should be AWSModel

File: docs/api/language_model_clients/aws.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+**Parameters:**
+- `aws_provider` (AWSProvider): The aws provider to use. One of `Bedrock` or `Sagemaker`.
+- `model` (_str_): Mistral AI pretrained models. Defaults to `mistral-medium-latest`.
+- `max_context_size` (_Optional[int]_, _optional_): Max context size for this model. Defaults to 32768.
+- `max_new_tokens` (_Optional[int]_, _optional_): Max new tokens possible for this model. Defaults to 1500.
+- `**kwargs`: Additional language model arguments to pass to the API provider.
+
+### Methods
+
+Refer to [`dspy.OpenAI`](https://dspy-docs.vercel.app/api/language_model_clients/OpenAI) documentation.
```
Comment (arnavsinghvi11) [inline docs/api/language_model_clients/aws.md:58]: it would be good to replace this with some commentary on the AWSMistral methods since it inherits from AWSModel, which has core functionality differing from LM

File: dsp/modules/aws_models.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+                "role": "user",
+                "content": [
+                    {
+                        "type": "text",
+                        "text": prompt,
+                    },
+                ],
+            },
+        ]
+        query_args["anthropic_version"] = "bedrock-2023-05-31"
```
Comment (arnavsinghvi11) [inline dsp/modules/aws_models.py:236]: should this be set through the kwargs?

File: dsp/modules/aws_models.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+CHARS2TOKENS: int = 4
+
+
+class AWSModel(LM):
+    """This class adds support for an AWS model.
+    It is an abstract class and should not be instantiated directly.
+    Instead, use one of the subclasses - AWSMistral, AWSAnthropic, or AWSMeta.
+    The subclasses implement the abstract methods _create_body and _call_model and work in conjunction with the AWSProvider classes Bedrock and Sagemaker.
+    Usage Example:
+        bedrock = Bedrock(region_name="us-west-2")
```
Comment (arnavsinghvi11) [inline dsp/modules/aws_models.py:24]: same comment as above

File: dsp/modules/aws_providers.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+from abc import ABC, abstractmethod
+from typing import Any, Optional
+
+
+class AWSProvider(ABC):
+    """This abstract class adds support for AWS model providers such as Bedrock and SageMaker.
+    The subclasses such as Bedrock and Sagemaker implement the abstract method _call_model and work in conjunction with the AWSModel classes.
+    Usage Example:
+        bedrock = Bedrock(region_name="us-west-2")
```
Comment (arnavsinghvi11) [inline dsp/modules/aws_providers.py:11]: same comment as above with dspy calls

File: dsp/modules/aws_providers.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        """_summary_.
+
+        Args:
+            region_name (str, optional): The AWS region where this LM is hosted.
+            service_name (str): Used in context of invoking the boto3 API.
+            profile_name (str, optional): boto3 credentials profile.
+            batch_n_enabled (bool): If False, call the LM N times rather than batching.
+        """
+        try:
+            import boto3  # pylint: disable=import-outside-toplevel
```
Comment (arnavsinghvi11) [inline dsp/modules/aws_providers.py:36]: remove comment

File: dsp/modules/aws_providers.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        except ImportError as exc:
+            raise ImportError('Please install boto3 to use AWS models.') from exc
```
Comment (arnavsinghvi11) [inline dsp/modules/aws_providers.py:38]: rephrase this as `pip install boto3`

**Comment by arnavsinghvi11** [comment]
Thanks @drawal1 ! looks great! just left a few minor comments but should be good to merge after that.

File: docs/api/language_model_clients/aws.md (hunk 1 above)
Comment (drawal1) [reply docs/api/language_model_clients/aws.md]: done!

File: docs/api/language_model_clients/aws.md (hunk 3 above)
Comment (drawal1) [reply docs/api/language_model_clients/aws.md]: done!

File: dsp/modules/aws_models.py (hunk 4 above)
Comment (drawal1) [reply dsp/modules/aws_models.py]: Good catch! I moved this to the constructor instead and made it conditional to Bedrock. Users can override. 
Ideally, we can put such params in a config file. Do we have support config files?

**Comment by drawal1** [comment]
@arnavsinghvi11 - Should be ready to merge now, hopefully :)

**Comment by arnavsinghvi11** [comment]
Thanks @drawal1 ! looks great now. Just a small thing before I merge. Could you add some further documentation for AWSProvider? Missed flagging this in earlier comments but something similar to what's done for AWSMistral and add it for an example AWSProvider like Bedrock. Would also be good to add an example of the AWS LM workflow to the documentation (connecting to provider, provider->model, configuring model in DSPy).

**Comment by drawal1** [comment]
@arnavsinghvi11 - AWSProvider is now documented as well

**Comment by arnavsinghvi11** [comment]
Thanks for all the contributions @drawal1!

**Comment by mikeusru** [comment]
> num_generations

@drawal1 it seems like num_generations is being set to None automatically if "n" and "num_generations" doesn't exist in kwargs, which causes the error in this line:
`if (temperature is None or temperature <= 0.15) and num_generations > 1:`

One potential fix is changing this to 
`if (temperature is None or temperature <= 0.15) and num_generations is not None and num_generations > 1:`

**Comment by drawal1** [comment]
@mikeusru - what aws provider and model are you using? can you share a code snippet so I can repro?

**Comment by mikeusru** [comment]
Using Bedrock and claude3-haiku

```
bedrock = dspy.Bedrock(...)

haiku = dspy.AWSAnthropic(
    aws_provider=bedrock,
    model="anthropic.claude-3-haiku-20240307-v1:0",
)

class BasicQA(dspy.Signature):
    """Answer questions with short factoid answers."""

    question = dspy.InputField()
    answer = dspy.OutputField(desc="often between 1 and 5 words")
    
# Define the predictor. Notice we're just changing the class. The signature BasicQA is unchanged.
generate_answer_with_chain_of_thought = dspy.ChainOfThought(BasicQA)

# Call the predictor on the same input.
pred = generate_answer_with_chain_of_thought(question=dev_example.question)

```

Error: `TypeError: '>' not supported between instances of 'NoneType' and 'int'`

The error is fixed when the change I posted earlier is introduced

**Comment by drawal1** [comment]
@mikeusru - good catch! I agree with your solution. Do you want to submit a PR? Otherwise, I can do it

**Comment by mikeusru** [comment]
> @mikeusru - good catch! I agree with your solution. Do you want to submit a PR? Otherwise, I can do it

I suppose I can submit the PR - my first ever :D
